### PR TITLE
Move GitHub tests to their own CI job (workaround for failing tests in github 0.36.5+)

### PR DESCRIPTION
**Pull request by DeeDeeG**
<details><summary>Requirements for Contributing a Bug Fix (from template, click to expand):</summary>

* Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* The pull request must only fix an existing bug. To contribute other changes, you must use a different template. You can see all templates at <https://github.com/atom/.github/tree/master/.github/PULL_REQUEST_TEMPLATE>.
* The pull request must update the test suite to demonstrate the changed functionality. For guidance, please see <https://flight-manual.atom.io/hacking-atom/sections/writing-specs/>.
* After you create the pull request, all status checks must be pass before a maintainer reviews your contribution. For more details, please see <https://github.com/atom/.github/tree/master/CONTRIBUTING.md#pull-requests>.

</details>

### Identify the Bug

Along with https://github.com/atom/github/pull/2617, unblocks https://github.com/atom/atom/pull/21782 _(also known as https://github.com/atom/github/issues/2607)_.

### Description of the Change

Move the `github` package tests out to their own CI job. This job manually runs `Atom --test` on the `github` package's test suite. It does so outside of `script/test`, which appears to avoid and work around the mysterious bug that's blocking progress in https://github.com/atom/atom/pull/21782.

Adds an env var `ATOM_PACKAGES_TO_SKIP` to `script/test`, allowing us to skip testing arbitrary packages by name -- This is used so that the `packages-1` test job doesn't attempt to run the `github` package tests, since it's being handled in the other job now. _(This is similar to the existing `ATOM_PACKAGES_TO_TEST` env var, and it does not conflict with any of the existing env vars or options to `script/test`. They all work together just fine.)_

There is another workaround required to get this to unblock https://github.com/atom/atom/pull/21782: Some other tests appear (perhaps unrelatedly?) to be failing under the new testing method. The additional workaround for that problem is in https://github.com/atom/github/pull/2617.

### Alternate Designs

- Stop running the `github` package tests here, and rely exclusively on the `atom/github` repo's CI
  - Disadvantageous, because if this repo breaks `github`'s CI, it could be caught much later, potentially after multiple unrelated PRs have been merged here, making bisecting issues very difficult. Could make maintaining both the `github` package and Atom core more fraught than need be.
- Discover the underlying issue and fix it rather than work around it
  - Multiple people have been stuck on this for weeks, so I don't think this will happen soon...

### Possible Drawbacks

In https://github.com/atom/github/pull/2617, I propose to disable some tests that fail only in this repository's CI. The same tests pass over at the `github` package's own CI. So we begin to rely on the atom/github repo's CI a bit more heavily than before.

### Verification Process

Many CI runs: https://dev.azure.com/DeeDeeG/b/_build?definitionId=17&branchFilter=129%2C144%2C146
- https://github.com/DeeDeeG/atom/commits/CI-separate-github-tests-job
- https://github.com/DeeDeeG/atom/commits/separate-github-tests-job

### Release Notes

N/A

### CODE REVIEW

**Comment by DeeDeeG** [comment]
@sadick254 if a new release of the`github` package gets published, this PR should allow test to pass for that new release being merged into Atom.

**Review by aminya — CHANGES_REQUESTED** [review]
I don't like this approach. We should not trick ourselves out. There is nothing special about GitHub tests. If they fail, they should be fixed in their own repository.

The PR for fixing GitHub CI is left unmerged simply because it brings the failures of the GitHub tests to the surface
https://github.com/atom/github/pull/2459

**Comment by DeeDeeG** [comment]
> The PR for fixing GitHub CI is left unmerged simply because it brings the failures of the GitHub tests to the surface
atom/github#2459

I do not agree with this statement. It is unmerged because it causes _new_ timeouts to surface. I do not believe those are related to the CI failure over at this repository. More tests could confirm whether your hypothesis or mine is correct:

- Test https://github.com/atom/github/pull/2459 against the `github` 0.36.3 release. If it fails for `0.36.3`, then the evidence is poorer that the errors you surfaced in https://github.com/atom/github/pull/2459 are related to the CI issues we saw here with `github` 0.36.5+.
- Test the `github` package again (with my workaround  https://github.com/atom/github/pull/2617 merged) under `script/test` again, trying without the present PR https://github.com/atom/atom/pull/21899. If it passes, then this PR is moot and it can be closed. I believe I already tested something like this, but I can try it again and report back.

<details><summary>Thoughts about the non-technical "philosophical" bits (click to expand if you want)</summary>

> I don't like this approach.

I happen to agree with you. But it's the only thing I've tried that works to test `github` at this repo, and I have tried many, many things. It seems the developer time to track down this bug, if it's worth tracking down (I don't think it's worth tracking down, at this rate), simply does not exist for the foreseeable future.

> There is nothing special about GitHub tests.

Yes. Except that it's a very prominent feature of Atom, lots of users use it and complain when it's broken, and right now the `0.36.3` release in stable is deleting people's git configs. The very PR with the fix for this important user-facing issue also broke Atom's tests in a bizarre and mysterious way that is hard to debug.

> If they fail, they should be fixed in their own repository.

I agree as a general principle. This is an unusual case. I am calling for an exception to this typical best practice, simply because the "best practice" path to resolve this issue has hit a dead end, and we are stuck worrying about a small test failure that I believe is our testing script's fault and or CI's fault -- I am willing to take the risk that this is not an issue in the real world.

</details>

File: script/vsts/platforms/templates/macos-tests-github.yml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+          $(Build.SourcesDirectory)/apm/node_modules/.bin/apm install
+        displayName: Prepare test dependencies for package
+        condition: and(succeeded(), ne(variables['Atom.SkipTests'], 'true'))
+
+      - pwsh: |
+          echo "CI env var is set to: $env:CI"
+          echo "MOCHA_TIMOUT env var is set to: $env:MOCHA_TIMEOUT"
+          echo "UNTIL_TIMOUT env var is set to: $env:UNTIL_TIMEOUT"
+
+          # Run tests, retrying up to 6 times if any tests fail
```
[the remark this answers is not in the document]
Comment (DeeDeeG) [reply script/vsts/platforms/templates/macos-tests-github.yml:42]: Haha, yeah. Um, this is really just to loosely imitate `script/test`, which has rerunning for specific timeout strings built in.

In my experience, for `github`, this retrying either did not help at all (even using 60-second timouts) or tests (taht I didn't disable somewhwre) reliably passed the first time.

Meaning: This really could be taken out, I think.

[the remark this answers is not in the document]
Comment (DeeDeeG) [reply script/vsts/platforms/templates/macos-tests-github.yml:42]: I would expect real flakes/timeouts maybe once a week or once a month? Or less often? Extrapolating from several runs where (when the `integration: file-patching` tests were disabled) this didn't fail once.

**Comment by smashwilson** [comment]
> > If they fail, they should be fixed in their own repository.
> 
> I agree as a general principle. This is an unusual case. I am calling for an exception to this typical best practice, simply because the "best practice" path to resolve this issue has hit a dead end, and we are stuck worrying about a small test failure that I believe is our testing script's fault and or CI's fault -- I am willing to take the risk that this is not an issue in the real world.

100%. My priority right now is to be able to ship bugfixes to Atom users in a scale of time faster than "years". I have reason to believe that the test failures that were holding us up are _not_ faults that manifest to end users, but instead are artifacts of things like quirks of CI VM environments.

@aminya, if you have the time to chase down:

1. Why the test process terminates unexpectedly halfway through printing a test case name, but only when running as a child process of `script/test` in this repository, _and_
2. How to improve the robustness of the rudimentary integration test suites in that package,

I will happily merge those pull requests!

File: script/vsts/platforms/templates/macos-tests-github.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (smashwilson) [reply script/vsts/platforms/templates/macos-tests-github.yml:42]: Maybe we could leave it in for a release cycle or so, then examine how often it triggered and was helpful? I'm not sure how easy it is to query for strings in CI build output like that, though. 🤔

**Comment by UziTech** [comment]
> My priority right now is to be able to ship bugfixes to Atom users in a scale of time faster than "years".

Than the github package should be a community package instead of a core package so users can get updates without needing to wait for a new release of Atom.

**Comment by DeeDeeG** [comment]
@UziTech can't folks do `apm install github`?

This is actually a bit documented now, though I admit it hasn't been proactively/widely surfaced or advertised: https://github.com/atom/github/blob/master/docs/installation.md#live-on-the-edge

**Comment by DeeDeeG** [comment]
Getting a bit off topic IMO:

> Then the github package should be a community package instead of a core package so users can get updates without needing to wait for a new release of Atom.

In the spirit of trying everything: I think we were at an impasse last time this was discussed of how to accomplish shipping Atom with a bundled community package. I believe somewhere in this repo there is a "skeleton" of a default `.atom` folder. I suppose the skeleton could be populated with the built `github` package, or whatever arbitrary package you want. (No promises it's easy to implement that, if possible at all. It just came to mind suddenly.)

If you want to follow up on that, let's make sure there is a separate appropriate issue for that, because it's not relevant to making the most-recent `github` tests pass. The whole point of the current PR is to arrive at **_passing tests_** that are still as useful as possible.

**Comment by aminya** [comment]
> Why the test process terminates unexpectedly halfway through printing a test case name, but only when running as a child process of script/test in this repository, and

We can make each package's tests more independent by spawning separate shells (and if required using `detached` option).

> How to improve the robustness of the rudimentary integration test suites in that package,

The tests should be written looser. This usually happens when you have assumptions about the environment that are not true. For example, in #21811, the assumption was that the scroll numbers are "exact". While in real UI tests, that does not happen.

**Comment by DeeDeeG** [comment]
UPDATE: I got exactly two passing run so far (out of two tried, 100% pass rate with two data points) of Atom CI... _without this present PR_, just using the latest workaround in the `github` package repo: https://github.com/atom/github/pull/2617

_(Looks good but I want to run it a few more times.)_

If that keeps passing, I'm going to close this as not needed.

**_PLEASE DO NOT MERGE, INTIAL RESULTS SHOW THIS PR IS MOST LIKELY NOT NEEDED AS OF https://github.com/atom/github/pull/2617_**

**Comment by DeeDeeG** [comment]
Tests are now four for four (100% passing) on my branch simply bumping to the updated `github` package on Atom's `master`. So... More tests on the way, but I am closing this now.

CLOSED as not needed, but please consider making a newer release of the `github` package at some point and merging that to Atom.

**Comment by DeeDeeG** [comment]
For context, the underlying problem was discovered in the comments of https://github.com/atom/github/pull/2459.

Some tests in `github` as of 0.36.5 assume `user.name` and `user.email` are configured for the `git` on the system PATH, and if those configs aren't set, the tests never complete.

**Pull request closed without merging**